File: 10_flutter_and_native_samples/ios/Runner/WifiStatusApiImpl.swift
import Foundation
import Network

/// Implements the Pigeon-generated `WifiStatusApi` using the Network framework.
///
/// The WifiStatus type is defined once in the Pigeon schema; this only fills it in.
final class WifiStatusApiImpl: WifiStatusApi {
  private let monitorQueue = DispatchQueue(label: "examples.wifi_status.monitor")

  func getWifiStatus(completion: @escaping (Result<WifiStatus, Error>) -> Void) {
    let monitor = NWPathMonitor()

    monitor.pathUpdateHandler = { path in
      let usesWifi = path.usesInterfaceType(.wifi)
      let satisfied = path.status == .satisfied

      let connectionType: WifiConnectionType
      if usesWifi {
        connectionType = .wifi
      } else if satisfied {
        connectionType = .other
      } else {
        connectionType = WifiConnectionType.none
      }

      // iOS has no public API for the radio toggle, so isEnabled is best effort.
      let status = WifiStatus(
        isEnabled: usesWifi,
        isConnected: usesWifi && satisfied,
        connectionType: connectionType,
        ssid: nil,
        signalLevel: nil
      )

      monitor.cancel()
      DispatchQueue.main.async {
        completion(.success(status))
      }
    }

    monitor.start(queue: monitorQueue)
  }
}

File: 10_flutter_and_native_samples/android/app/src/main/kotlin/com/example/flutter_and_native_samples/WifiStatusApiImpl.kt
package com.example.flutter_and_native_samples

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.wifi.WifiManager

/**
 * Implements the Pigeon-generated [WifiStatusApi] using Android system services.
 *
 * The WifiStatus type is defined once in the Pigeon schema; this only fills it in.
 */
class WifiStatusApiImpl(context: Context) : WifiStatusApi {
    private val appContext = context.applicationContext

    override fun getWifiStatus(callback: (Result<WifiStatus>) -> Unit) {
        try {
            val wifiManager =
                appContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            val connectivityManager =
                appContext.getSystemService(Context.CONNECTIVITY_SERVICE)
                    as ConnectivityManager

            val capabilities =
                connectivityManager.getNetworkCapabilities(
                    connectivityManager.activeNetwork,
                )
            val usesWifi =
                capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true

            val connectionType =
                when {
                    usesWifi -> WifiConnectionType.WIFI
                    capabilities != null -> WifiConnectionType.OTHER
                    else -> WifiConnectionType.NONE
                }

            callback(
                Result.success(
                    WifiStatus(
                        isEnabled = wifiManager.isWifiEnabled,
                        isConnected = usesWifi,
                        connectionType = connectionType,
                        ssid = null,
                        signalLevel = null,
                    ),
                ),
            )
        } catch (error: Exception) {
            callback(Result.failure(error))
        }
    }
}
